### PR TITLE
Add quasiquoters that can handle C++ exceptions.

### PR DESCRIPTION
**Pull request by chpatrick**

### CODE REVIEW

File: inline-c-cpp/src/Language/C/Inline/Cpp/Exceptions.hs [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+import           Control.Exception.Safe
+import           Control.Monad
+import qualified Language.C.Inline as C
+import           Language.Haskell.TH
+import           Language.Haskell.TH.Quote
+import           Foreign
+import           Foreign.C
+
+-- | An exception thrown in C++ code, with the message from @what()@.
+newtype ForeignException = ForeignException String
```
Comment (bitonic) [inline inline-c-cpp/src/Language/C/Inline/Cpp/Exceptions.hs]: I'd call this `CppException`, `ForeignException` is a bit generic.

File: inline-c-cpp/src/Language/C/Inline/Cpp/Exceptions.hs [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+            , "    std::memcpy(message, e.what(), whatLen);"
+            , "    *$(const char** " ++ nameBase errPtrVarName ++") = message;"
+            , "  } catch (...) {"
+            , "    char* message = static_cast<char*>(std::malloc(1));"
+            , "    *message = 0;"
+            , "    *$(const char** " ++ nameBase errPtrVarName ++") = message;"
+            , "  }"
+            , "}"
+            ]
+      [e| handleForeign $ \ $(varP errPtrVarName) -> $(quoteExp C.exp inlineCStr) |]
```
Comment (bitonic) [inline inline-c-cpp/src/Language/C/Inline/Cpp/Exceptions.hs]: shouldn't this be a `C.block`?

File: inline-c-cpp/src/Language/C/Inline/Cpp/Exceptions.hs (hunk 1 above)
Comment (bitonic) [inline inline-c-cpp/src/Language/C/Inline/Cpp/Exceptions.hs]: would it be possible to distinguish between `std::exception`s and non-exceptions here? e.g. with a sum type in `CppException`.

Comment (chpatrick) [inline inline-c-cpp/src/Language/C/Inline/Cpp/Exceptions.hs]: Done.

File: inline-c-cpp/src/Language/C/Inline/Cpp/Exceptions.hs (hunk 2 above)
Comment (chpatrick) [inline inline-c-cpp/src/Language/C/Inline/Cpp/Exceptions.hs]: Done.

File: inline-c-cpp/src/Language/C/Inline/Cpp/Exceptions.hs (hunk 1 above)
Comment (chpatrick) [inline inline-c-cpp/src/Language/C/Inline/Cpp/Exceptions.hs]: Done.

**Comment by bitonic** [comment]
Reviewed 2 of 4 files at r1, 2 of 2 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/fpco/inline-c/59)*

**Pull request merged**